File: Pynq-Z1/vivado/ip/pmod_io_switch_1.0/drivers/pmod_io_switch_v1_0/src/pmod.c
/******************************************************************************
 *  Copyright (c) 2016, Xilinx, Inc.
 *  All rights reserved.
 * 
 *  Redistribution and use in source and binary forms, with or without 
 *  modification, are permitted provided that the following conditions are met:
 *
 *  1.  Redistributions of source code must retain the above copyright notice, 
 *     this list of conditions and the following disclaimer.
 *
 *  2.  Redistributions in binary form must reproduce the above copyright 
 *      notice, this list of conditions and the following disclaimer in the 
 *      documentation and/or other materials provided with the distribution.
 *
 *  3.  Neither the name of the copyright holder nor the names of its 
 *      contributors may be used to endorse or promote products derived from 
 *      this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, 
 *  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
 *  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
 *  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, 
 *  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
 *  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 *  OR BUSINESS INTERRUPTION). HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
 *  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR 
 *  OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF 
 *  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *****************************************************************************/
/******************************************************************************
 *
 *
 * @file pmod.c
 *
 * Implementing useful functions, including the IIC read and write, 
 * initialization functions for SPI devices, delay functions, timer setup, 
 * etc.
 *
 * <pre>
 * MODIFICATION HISTORY:
 *
 * Ver   Who  Date     Changes
 * ----- --- ------- -----------------------------------------------
 * 1.00a pp  04/29/16 release
 * 1.00a pp  07/05/16 updated to be consistent with arduino io switch driver
 * </pre>
 *
 *****************************************************************************/

#include "pmod.h"

// The Timer Counter instance
XTmrCtr TimerInst_0;

void spi_transfer(u32 BaseAddress, int bytecount, 
                    u8* readBuffer, u8* writeBuffer) {
    int i;

    XSpi_WriteReg(BaseAddress,XSP_SSR_OFFSET, 0xfe); // 0xfe
    for (i=0; i<bytecount; i++)
    {
        XSpi_WriteReg(BaseAddress,XSP_DTR_OFFSET, writeBuffer[i]);
    }
    while(((XSpi_ReadReg(BaseAddress,XSP_SR_OFFSET) & 0x04)) != 0x04);
    // delay for about 100 ns
    XTmrCtr_SetResetValue(&TimerInst_0, 1, 10);
    // Start the timer0
    XTmrCtr_Start(&TimerInst_0, 1);
    // Wait for the delay to lapse
    while(!XTmrCtr_IsExpired(&TimerInst_0,1));
    // Stop the timer0
    XTmrCtr_Stop(&TimerInst_0, 1);
    
    // Read SPI
    for(i=0;i< bytecount; i++){
       readBuffer[i] = XSpi_ReadReg(BaseAddress,XSP_DRR_OFFSET);
    }
    XSpi_WriteReg(BaseAddress, XSP_SSR_OFFSET, 0xff);
}

void spi_init(u32 BaseAddress, u32 clk_phase, u32 clk_polarity){
    u32 Control;

    // Soft reset SPI
    XSpi_WriteReg(BaseAddress, XSP_SRR_OFFSET, 0xa);
    // Master mode
    Control = XSpi_ReadReg(BaseAddress, XSP_CR_OFFSET);
    // Master Mode
    Control |= XSP_CR_MASTER_MODE_MASK;
    // Enable SPI
    Control |= XSP_CR_ENABLE_MASK;
    // Slave select manually
    Control |= XSP_INTR_SLAVE_MODE_MASK;
    // Enable Transmitter
    Control &= ~XSP_CR_TRANS_INHIBIT_MASK;
    // XSP_CR_CLK_PHASE_MASK
    if(clk_phase)
    	Control |= XSP_CR_CLK_PHASE_MASK;
    // XSP_CR_CLK_POLARITY_MASK
    if(clk_polarity)
    	Control |= XSP_CR_CLK_POLARITY_MASK;
    XSpi_WriteReg(BaseAddress, XSP_CR_OFFSET, Control);
}

int iic_read(u32 iic_BaseAddress, u32 addr, u8* buffer, u8 numbytes){
    XIic_Recv(iic_BaseAddress, addr, buffer, numbytes, XIIC_STOP);
    return 0;
}


int iic_write(u32 iic_BaseAddress, u32 addr, u8* buffer, u8 numbytes){
       XIic_Send(iic_BaseAddress, addr, buffer, numbytes, XIIC_STOP);
       return 0;
}

int cb_init(circular_buffer *cb, volatile u32* log_start_addr, 
            size_t capacity, size_t sz){
  cb->buffer = (volatile char*) log_start_addr;
  if(cb->buffer == NULL)
    return -1;
  cb->buffer_end = (char *)cb->buffer + capacity * sz;
  cb->capacity = capacity;
  cb->sz = sz;
  cb->head = cb->buffer;
  cb->tail = cb->buffer; 

  // Mailbox API Initialization
  MAILBOX_DATA(0)  = 0xffffffff;
  MAILBOX_DATA(2)  = (u32) cb->head;
  MAILBOX_DATA(3)  = (u32) cb->tail;
  
  return 0;
}

void cb_push_back(circular_buffer *cb, const void *item){

  u8 i;
  u8* tail_ptr = (u8*) cb->tail;
  u8* item_ptr = (u8*) item;

  // update data 
  for(i=0;i<cb->sz;i++){
    tail_ptr[i] = item_ptr[i]; 
  }

  cb_push_incr_ptrs(cb);

  // Mailbox API Update
  MAILBOX_DATA(0)  = (u32) item;
  MAILBOX_DATA(2)  = (u32) cb->head;
  MAILBOX_DATA(3)  = (u32) cb->tail;
}

void cb_push_back_float(circular_buffer *cb, const float *item){
  // update data 
  float* tail_ptr = (float*) cb->tail;
  *tail_ptr = *item;
  
  cb_push_incr_ptrs(cb);

  // Mailbox API Update
  MAILBOX_DATA_FLOAT(0)  = *item;
}

void cb_push_incr_ptrs(circular_buffer *cb){

  // Update pointers
  cb->tail = (char*)cb->tail + cb->sz;
  if(cb->tail >= cb->buffer_end)
    cb->tail = cb->buffer;

  if((cb->tail == cb->head) ) {
    cb->head  = (char*)cb->head + cb->sz;
  }

  // Update mailbox API
  MAILBOX_DATA(2) = (u32) cb->head;
  MAILBOX_DATA(3) = (u32) cb->tail;
}

void delay_us(int usdelay){
    // us delay
    XTmrCtr_SetResetValue(&TimerInst_0, 1, usdelay*100);
    // Start the timer0 for usdelay us delay
    XTmrCtr_Start(&TimerInst_0, 1);
    // Wait for usdelay us to lapse
    while(!XTmrCtr_IsExpired(&TimerInst_0,1));
    // Stop the timer0
    XTmrCtr_Stop(&TimerInst_0, 1);
}

void delay_ms(u32 msdelay){
    // ms delay
    XTmrCtr_SetResetValue(&TimerInst_0, 1, msdelay*100*1000);
    // Start the timer0 for usdelay us delay
    XTmrCtr_Start(&TimerInst_0, 1);
    // Wait for usdelay us to lapse
    while(!XTmrCtr_IsExpired(&TimerInst_0,1));
    // Stop the timer0
    XTmrCtr_Stop(&TimerInst_0, 1);
}

/*
 * Switch Configuration
 * 8 input chars, each representing the connection for 1 Pmod pin
 * Only the least significant 4-bits for each input char is used
 *
 * Configuration is done by writing a 32 bit value to the switch.
 * The 32-bit value represents 8 x 4-bit values concatenated.
 * One 4-bit value is used to configure each Pmod pin.
 * Pmod pin 8 = bits [31:28]
 * Pmod pin 7 = bits [27:24]
 * Pmod pin 6 = bits [23:20]
 * Pmod pin 5 = bits [19:16]
 * Pmod pin 4 = bits [15:12]
 * Pmod pin 3 = bits [11:8]
 * Pmod pin 2 = bits [7:4]
 * Pmod pin 1 = bits [3:0]
 * e.g. Write GPIO 0 - 7 to Pmod 1-8 => switchConfigValue = 0x76543210
 */
void config_pmod_switch(char pin0, char pin1, char pin2, char pin3, 
                        char pin4, char pin5, char pin6, char pin7){
   u32 switchConfigValue;

   // Calculate switch configuration value
   switchConfigValue = (pin7<<28)|(pin6<<24)|(pin5<<20)|(pin4<<16)|\
                        (pin3<<12)|(pin2<<8)|(pin1<<4)|(pin0);

   // isolate switch by writing 0 to bit 31
   Xil_Out32(SWITCH_BASEADDR+0x4,0x00000000);
   // Set pin configuration
   Xil_Out32(SWITCH_BASEADDR, switchConfigValue);
   // Re-enable Swtch by writing 1 to bit 31
   Xil_Out32(SWITCH_BASEADDR+0x4,0x80000000);
}

int tmrctr_init(void) {
    int Status;

    // specify the device ID that is generated in xparameters.h
    Status = XTmrCtr_Initialize(&TimerInst_0, XPAR_TMRCTR_0_DEVICE_ID);
    if (Status != XST_SUCCESS) {
        return XST_FAILURE;
    }

    XTmrCtr_SetOptions(&TimerInst_0, 1, XTC_AUTO_RELOAD_OPTION | \
                            XTC_CSR_LOAD_MASK | XTC_CSR_DOWN_COUNT_MASK);
    XTmrCtr_Start(&TimerInst_0, 1);

    return 0;
}

void pmod_init(u32 clk_phase, u32 clk_polarity) {
    spi_init(SPI_BASEADDR, clk_phase, clk_polarity);
    tmrctr_init();
}

File: pynq/lib/pmod/pmod.py
#   Copyright (c) 2016, Xilinx, Inc.
#   All rights reserved.
#
#   Redistribution and use in source and binary forms, with or without
#   modification, are permitted provided that the following conditions are met:
#
#   1.  Redistributions of source code must retain the above copyright notice,
#       this list of conditions and the following disclaimer.
#
#   2.  Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#
#   3.  Neither the name of the copyright holder nor the names of its
#       contributors may be used to endorse or promote products derived from
#       this software without specific prior written permission.
#
#   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
#   AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
#   THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
#   PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
#   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
#   EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
#   PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
#   OR BUSINESS INTERRUPTION). HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
#   WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
#   OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
#   ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import asyncio
import os
import sys
import math
from pynq.lib import PynqMicroblaze
from pynq.lib.pynqmicroblaze import add_bsp
from . import MAILBOX_OFFSET
from . import MAILBOX_PY2IOP_CMD_OFFSET
from . import BIN_LOCATION
from . import BSP_LOCATION

__author__ = "Yun Rock Qu"
__copyright__ = "Copyright 2016, Xilinx"
__email__ = "yunq@xilinx.com"


class Pmod(PynqMicroblaze):
    """This class controls the Pmod Microblaze instances in the system.

    This class inherits from the PynqMicroblaze class. It extends 
    PynqMicroblaze with capability to control Pmod devices.

    Attributes
    ----------
    ip_name : str
        The name of the IP corresponding to the Microblaze.
    rst_name : str
        The name of the reset pin for the Microblaze.
    mb_program : str
        The absolute path of the Microblaze program.
    state : str
        The status (IDLE, RUNNING, or STOPPED) of the Microblaze.
    reset_pin : GPIO
        The reset pin associated with the Microblaze.
    mmio : MMIO
        The MMIO instance associated with the Microblaze.
    interrupt : Event
        An asyncio.Event-like class for waiting on and clearing interrupts.

    """

    def __init__(self, mb_info, mb_program):
        """Create a new Microblaze object.

        This method leverages the initialization method of its parent. It 
        also deals with relative / absolute path of the program.

        Parameters
        ----------
        mb_info : dict
            A dictionary storing Microblaze information, such as the 
            IP name and the reset name.
        mb_program : str
            The Microblaze program loaded for the processor.

        Examples
        --------
        The `mb_info` is a dictionary storing Microblaze information:

        >>> mb_info = {'ip_name': 'mb_bram_ctrl_1',
        'rst_name': 'mb_reset_1', 
        'intr_pin_name': 'iop1/dff_en_reset_0/q', 
        'intr_ack_name': 'mb_1_intr_ack'}

        """
        if not os.path.isabs(mb_program):
            mb_program = os.path.join(BIN_LOCATION, mb_program)

        super().__init__(mb_info, mb_program)

    def write_mailbox(self, data_offset, data):
        """This method write data into the mailbox of the Microblaze.

        Parameters
        ----------
        data_offset : int
            The offset for mailbox data, 0,4,... for MAILBOX 0,1,...
        data : int/list
            A list of 32b words to be written into the mailbox.

        Returns
        -------
        None

        """
        offset = MAILBOX_OFFSET + data_offset
        self.write(offset, data)

    def read_mailbox(self, data_offset, num_words=1):
        """This method reads mailbox data from the Microblaze.

        Parameters
        ----------
        data_offset : int
            The offset for mailbox data, 0,4,... for MAILBOX 0,1,...
        num_words : int
            Number of 32b words to read from Microblaze mailbox.

        Returns
        -------
        int/list
            An int of a list of data read from the mailbox.

        """
        offset = MAILBOX_OFFSET + data_offset
        return self.read(offset, num_words)

    def write_blocking_command(self, command):
        """This method writes a blocking command to the Microblaze.

        The program waits in the loop until the command is cleared by the
        Microblaze.

        Parameters
        ----------
        command : int
            The command to write to the Microblaze.

        Returns
        -------
        None

        """
        self.write(MAILBOX_OFFSET + MAILBOX_PY2IOP_CMD_OFFSET, command)
        while self.read(MAILBOX_OFFSET + MAILBOX_PY2IOP_CMD_OFFSET) != 0:
            pass

    def write_non_blocking_command(self, command):
        """This method writes a non-blocking command to the Microblaze.

        The program will just send the command and returns the control 
        immediately.

        Parameters
        ----------
        command : int
            The command to write to the Microblaze.

        Returns
        -------
        None

        """
        self.write(MAILBOX_OFFSET + MAILBOX_PY2IOP_CMD_OFFSET, command)


if os.path.exists(BSP_LOCATION):
    add_bsp(BSP_LOCATION)
